internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,314 @@
+	setReplicas(s.machineSet, s.machines)
+
+	// Conditions
+
+	// Update the ScalingUp and ScalingDown condition, require the above setReplicas function.
+	setScalingUpCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound)
+	setScalingDownCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound, s.owningMachineDeployment)
+
+	// MachinesReady condition: aggregate the Machine's Ready condition
+	setMachinesReadyCondition(s.machineSet, s.machines)
+
+	// MachinesUpToDate condition: aggregate the Machine's UpToDate condition
+	setMachinesUpToDateCondition(s.machineSet, s.machines)
+
+	// TODO Deleting
+}
+
+func setReplicas(ms *clusterv1.MachineSet, machines []*clusterv1.Machine) {
+	// Return early when machines is nil. It's not possible to calculate replica counters.
+	// Conditions may surface to be unknown in this case.
+	if machines == nil {
+		return
+	}
+
+	var readyReplicas, availableReplicas int32
+	for _, machine := range machines {
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineReadyV1Beta2Condition) {
+			readyReplicas++
+		}
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineAvailableV1Beta2Condition) {
+			availableReplicas++
+		}
+	}
+
+	if ms.Status.V1Beta2 == nil {
+		ms.Status.V1Beta2 = &clusterv1.MachineSetV1Beta2Status{}
+	}
+
+	ms.Status.V1Beta2.ReadyReplicas = ptr.To(readyReplicas)
+	ms.Status.V1Beta2.AvailableReplicas = ptr.To(availableReplicas)
+}
+
+func setScalingUpCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingUpInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
+	// This could e.g. be the case when using autoscaling with MachineSets.
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetScalingUpWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+	if !ms.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+	gotReplicas := ms.Status.Replicas
+
+	referencesMessage := calculateMissingReferencesMessage(ms, bootstrapObjectNotFound, infrastructureObjectNotFound)
+
+	// Not scaling up or in deletion.
+	if gotReplicas >= desiredReplicas {
+		var message string
+		if referencesMessage != "" {
+			message = fmt.Sprintf("ScalingUp can't happen %s", referencesMessage)
```

Comment:
I think in this case we definitely have to add somethign that we also don't want to scale up. Otherwise it sounds like we dont' scale up only because the templates are missing, but actually we also don't scale up because there is no reason to (based on replica counters)

(would be easier to finalize once we have unit tests for all these cases)

Reply:
I'd say the NotScalingUp reason should be good enough to show this, wdyt?

Reply:
The problem is that the reason is not scaling up and if the message only contains the reference message it sounds like that message is the root cause of why we are not scaling up